src/UI/Actions.hs

```diff
@@ -287,44 +299,52 @@ instance HasList 'ListOfFiles where
 -- particular mail. To apply his changes, he 'completes' his typed in
 -- text by pressing Enter.
 --
-class Completable (m :: Name) where
-  complete :: Proxy m -> AppState -> T.EventM Name AppState
+class Completable (n :: Name) where
+  complete :: (MonadIO m, MonadState AppState m) => Proxy n -> m ()
 
 instance Completable 'SearchThreadsEditor where
   complete _ = applySearch
 
 instance Completable 'ManageMailTagsEditor where
-  complete _ s = liftIO $ over (asMailIndex . miMailTagsEditor . E.editContentsL) clearZipper <$> completeMailTags s
+  complete _ = do
+    get >>= liftIO . completeMailTags >>= put
+    modifying (asMailIndex . miMailTagsEditor . E.editContentsL) clearZipper
 
 instance Completable 'ComposeListOfAttachments where
-  complete _ = sendMail
+  complete _ = get >>= sendMail >>= put
 
 -- | Apply all given tag operations to existing mails
 --
 completeMailTags :: AppState -> IO AppState
 completeMailTags s =
     case getEditorTagOps (Proxy :: Proxy 'ManageMailTagsEditor) s of
         Left err -> pure $ setError err s
-        Right ops' -> over (asMailIndex . miListOfThreads) (L.listModify (Notmuch.tagItem ops'))
-                      <$> selectedItemHelper (asMailIndex . miListOfMails) s (manageMailTags s ops')
+        Right ops' -> flip execStateT s $ do
+          modifying (asMailIndex . miListOfThreads) (L.listModify (Notmuch.tagItem ops'))
+          selectedItemHelper (asMailIndex . miListOfMails) (manageMailTags ops')
+
+
+hide, unhide :: (MonadState AppState m) => ViewName -> Int -> Name -> m ()
```

Comment:
Ah nice. Well thank you! Although if it isn't too much effort I'd put them in `UI.Views`?

Reply:
not right now.  there's already enough in this refactor :D